Improve E2E Test Fixtures to be less flaky

JustinGrote opened the pull request:
- [x] Rewrite Pses Stdio host to be task-based and interface-based
- [x] Replace `LoggingStream` with `DebugOutputStream` to log LSP communication to debug console when debugging tests
- [x] Fix up Tests that require help by offline updating help in all platforms/CI

~~- [ ] Wire up LSP Client logging to XUnit `ITestOutputHelper`~~
~~- [ ] Wire up PSES Initial File Logging to debug console~~

JustinGrote commented:
@andyleejordan @SeeminglyScience Did a lot of cleanup here, DAP tests run much faster and I should have cleaned up the race conditions, they've been completely reliable for me so far. The tests should also read fairly cleaner now as well.

andyleejordan commented:
@JustinGrote running this through the internal pipeline right now!

andyleejordan commented:
> @JustinGrote running this through the internal pipeline right now!

Sooo actually per 20fa478dcd61bf75c84fe69f2f3f17ea4233d2e0 we cannot currently run the tests on OneBranch since it's stuck with PowerShell 7.3...so this will have to wait until January 😵

andyleejordan commented:
Hi @JustinGrote, we can _finally_ run the tests again in ADO. Would you want to get this PR updated? Note that I had to skip the `getCommand` test, it just hung (@SeeminglyScience's guess was that the ADO PowerShell might have network shares in `PSModulePath` but yeah it just ran for twenty minutes before I killed it).

JustinGrote commented:
@andyleejordan rebased and passed locally for me. Lets see if ADO likes it.
![image](https://github.com/user-attachments/assets/3c339cfe-bf79-4e14-80ea-18c452719aa7)

JustinGrote commented:
Looks like it works. I can do the logging fixups as a separate item especially now that the other logging stuff was implemented.

JustinGrote commented on test/PowerShellEditorServices.Test.E2E/DebugAdapterProtocolMessageTests.cs:
```diff
@@ -176,88 +175,114 @@ private string GenerateScriptFromLoggingStatements(params string[] logStatements
             {
-                await Task.Delay(1000);
+                try
+                {
+                    scriptLogReader = new StreamReader(
+                        new FileStream(
+                            testScriptLogPath,
+                            FileMode.OpenOrCreate,
+                            FileAccess.Read, // Because we use append, its OK to create the file ahead of the script
+                            FileShare.ReadWrite
+                        )
+                    );
+                }
+                catch (IOException) //Sadly there does not appear to be a xplat way to wait for file availability, but luckily this does not appear to fire often.
+                {
+                    await Task.Delay(500);
+                }
             }
-            // Sleep one more time after the file exists so whatever is writing can finish.
-            await Task.Delay(1000);
-            return File.ReadLines(s_testOutputPath).ToArray();
+
+            // return valid lines only
+            string nextLine = string.Empty;
+            while (nextLine is null || nextLine.Length == 0)
+            {
+                nextLine = await scriptLogReader.ReadLineAsync(); //Might return null if at EOF because we created it above but the script hasn't written to it yet
+            }
+            return nextLine;
```
Thanks copilot, but it's just a test.

JustinGrote commented on test/PowerShellEditorServices.Test.E2E/Hosts/PsesStdioLanguageServerProcessHost.cs:
```diff
@@ -0,0 +1,84 @@
+        s_binDir, $"pses_test_logs_{Path.GetRandomFileName()}");
+
+    private const string s_logLevel = "Diagnostic";
+    private static readonly string[] s_featureFlags = { "PSReadLine" };
```
I just looked and I didn't see this feature flag actually referenced anywhere, so it can probably be removed but I think that should be the scope of another PR.

JustinGrote commented on test/PowerShellEditorServices.Test.E2E/LanguageServerProtocolMessageTests.cs:
```diff
@@ -1078,24 +1085,22 @@ await PsesLanguageClient
         [SkippableFact]
         public async Task CanSendHoverRequestAsync()
         {
             Skip.If(IsLinux, "This depends on the help system, which is flaky on Linux.");
```
Maybe not? As you said lets YOLO it as a test.

JustinGrote replied:
I also think I separately fixed this test in the other branch so there might be a conflict, I'll rebase.

JustinGrote commented on test/PowerShellEditorServices.Test.E2E/LanguageServerProtocolMessageTests.cs:
```diff
@@ -1177,7 +1181,7 @@ await PsesLanguageClient
         {
-            Skip.If(PsesStdioProcess.RunningInConstrainedLanguageMode && PsesStdioProcess.IsWindowsPowerShell,
+            Skip.If(PsesStdioLanguageServerProcessHost.RunningInConstrainedLanguageMode && PsesStdioLanguageServerProcessHost.IsWindowsPowerShell,
                 "Windows PowerShell doesn't trust PSScriptAnalyzer by default so it won't load.");
```
It might be! Maybe I can get some LSP trace logs from them and stick in some telemetry to find out.

JustinGrote commented on test/PowerShellEditorServices.Test.E2E/PowerShellEditorServices.Test.E2E.csproj:
```diff
@@ -37,4 +40,11 @@
+
+  <ItemGroup>
+    <!-- Used for Update-Help for some test fixtures -->
+    <Content Include="..\PowerShellEditorServices.Test.Shared\PSHelp\**\*"
```
Only if the tests use Microsoft.PowerShell.Utility cmdlets for their reference. This is only required for Windows PowerShell since it doesn't ship this help in-box, all pwsh flavors do ship help in-box and so we don't have to do this. Assuming we find/adjust all help-based tests to use Microsoft.PowerShell.Utility, then yes this should fix them in AzDO CI

JustinGrote commented on test/PowerShellEditorServices.Test.E2E/LanguageServerProtocolMessageTests.cs:
```diff
@@ -1231,7 +1235,7 @@ await PsesLanguageClient
+        [Fact(Timeout = 60000)]
         public async Task CanSendGetCommandRequestAsync()
         {
             Skip.If(Environment.GetEnvironmentVariable("TF_BUILD") is not null,
```
It is super slow still in GH CI, since it is fetching *all* commands. I'll put it back into skip but we may want to see if it at least completes on non WinPS in ADO sometime.

andyleejordan commented:
LMAO Windows is mad now, sorry.

JustinGrote commented:
@andyleejordan  I'm not done fixing it, I'll flip it back to draft for now

JustinGrote commented:
@andyleejordan the `FindSymbolsInDSCFile` test failure appears to be because of 0135baaf8d6c118457867ac7481a74d067a2426a #1984. You updated the IDs of DSC symbols to include a prefix, but the test doesn't check for that new prefix.
https://github.com/PowerShell/PowerShellEditorServices/blob/de0eea18e8ce509915e8fdd8faeae0e2993c40c8/src/PowerShellEditorServices/Services/Symbols/Visitors/SymbolVisitor.cs#L275-L276

I will "fix" this test but please confirm if you're expecting something else. Also, completions allow for a `sortText` property in the LSP, maybe we should implement this more explicitly if that's the intention for the prefix?

JustinGrote commented:
@andyleejordan I came up with a better way to do the "offline" help update in eaeefa13f017d94bd9ad417ce6e72c406b29f874 and centralized it in the build script again.

andyleejordan commented:
> @andyleejordan the `FindSymbolsInDSCFile` test failure appears to be because of [0135baa](https://github.com/PowerShell/PowerShellEditorServices/commit/0135baaf8d6c118457867ac7481a74d067a2426a) #1984. You updated the IDs of DSC symbols to include a prefix, but the test doesn't check for that new prefix.
> 
> https://github.com/PowerShell/PowerShellEditorServices/blob/de0eea18e8ce509915e8fdd8faeae0e2993c40c8/src/PowerShellEditorServices/Services/Symbols/Visitors/SymbolVisitor.cs#L275-L276
> 
> I will "fix" this test but please confirm if you're expecting something else. Also, completions allow for a `sortText` property in the LSP, maybe we should implement this more explicitly if that's the intention for the prefix?

Oops. I don't think I'm expecting something else. And if I remember correctly on the sortText, VS Code completely ignored it and applies their own sorting.

JustinGrote commented:
@andyleejordan this should be ready for re-review. Changes summary:
1. Fixed help-based tests to all use `Microsoft.PowerShell.Archive`, because it has a small help file to be vendored into the codebase
1. Added update help build script for all pwsh platforms with verify check.
1. Added `upterm` to the CI process pre-build. This allows, if you explicitly re-run a job with debug build, to be able to SSH into the runner to interactively troubleshoot/debug issues. Only runs on debug re-runs
1. Most tests conditional on platforms fixed or mitigated.

andyleejordan reviewed (changes requested):
Sorry I was being picky about my build script lol, going to commit these, test, and merge!

andyleejordan commented:
That's what I get for doing this on GitHub.

andyleejordan commented:
It was because I added a single quote...inside single quotes 🫨

JustinGrote commented on .github/workflows/ci-test.yml:
```diff
@@ -51,6 +51,12 @@ jobs:
 
+      - name: If Debugging, start upterm for interactive pipeline troubleshooting
+        if: ${{ runner.debug == 1 }}
+        uses: lhotari/action-upterm@v1
```
I'm working on a version that lets you vscode remote into the github action, the tricky part is the authentication but it's definitely doable :)

JustinGrote commented:
@andyleejordan  thank you!
